### PR TITLE
Blockwise mask fn as opt arg in all masking functions

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

As per title, I think this pattern is used quite often and deserves to be a public mask-fn. Used currently in gemma/paligemma family, GIT, PI0 and will be used in two upcoming models (deepseekOcr and Molmo2)

This PR allows all these models ton go `non-vmap` path, which iirc is more preferable for us.

I opted for adding 'blockwise_mask' as an arg in existing mask fn, it doesn't seems like a new mask type of itself. We might have to create otherwise - `create_blockwise_causal_mask`, `create_blockwise_sliding_causal_mask`, etc.

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, gemma4, git, paligemma, pi0

**Comment by zucchini-nlp** [comment]
oh, no, I forgot about conversion. Will wait until merged and trigger test again

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, gemma4, git, paligemma, pi0

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, paligemma, pi0

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, paligemma, pi0

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, paligemma, pi0

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 1]
```diff
@@ -353,21 +254,30 @@ def forward(
...
+            "past_key_values": past_key_values,
+            "position_ids": position_ids,
+        }
+        is_first_iteration = past_key_values is None or not past_key_values.is_initialized or pixel_values is not None
+        if token_type_ids is not None and is_first_iteration:
+            # Can attend bidirectionally in prefix and only causally in suffix
+            mask_kwargs["block_sequence_ids"] = torch.where(token_type_ids == 0, 0, -1)
+
+        # PG has no sliding window, only full attn. But PG2 needs sliding mask and full mask
+        causal_mask = create_causal_mask(**mask_kwargs)
```
Comment (zucchini-nlp) [inline src/transformers/models/paligemma/modeling_paligemma.py:271]: aligns with normal LLMs that call `create_causal_mask`, not the generate-wrapper

File: src/transformers/masking_utils.py [hunk 2]
```diff
@@ -111,6 +111,32 @@ def inner_mask(batch_idx: int, head_idx: int, q_idx: int, kv_idx: int) -> bool:
...
+    def inner_mask(batch_idx: int, head_idx: int, q_idx: int, kv_idx: int) -> bool:
+        seq_length = block_sequence_ids.shape[-1]
+
+        # clamp indices because with static cache they can go beyond `block_sequence_ids.shape[-1]`
+        q_idx_clamped = q_idx.clamp(max=seq_length - 1)
+        kv_idx_clamped = kv_idx.clamp(max=seq_length - 1)
+
+        # Unmask if the q and kv come from same group which is not -1 (i.e. non-text)
+        q_group = block_sequence_ids[batch_idx, q_idx_clamped]
+        kv_group = block_sequence_ids[batch_idx, kv_idx_clamped]
```
Comment (Cyrilvallez) [inline src/transformers/masking_utils.py]: Instead of this, it's probably easier/better to simply pad `block_sequence_ids` before it enters here (same as we do with the `padding_mask`)

File: src/transformers/masking_utils.py [hunk 3]
```diff
@@ -943,6 +974,8 @@ def create_causal_mask(
 
     batch_size, dtype, device = inputs_embeds.shape[0], inputs_embeds.dtype, inputs_embeds.device
     mask_factory_function = causal_mask_function
+    if block_sequence_ids is not None:
+        mask_factory_function = or_masks(blockwise_overlay(block_sequence_ids), mask_factory_function)
```
Comment (Cyrilvallez) [inline src/transformers/masking_utils.py]: Let's do it immediately here instead of checking twice

```suggestion
    if block_sequence_ids is not None:
        mask_factory_function = or_masks(blockwise_overlay(block_sequence_ids), mask_factory_function)
        allow_is_causal_skip = False
```

File: src/transformers/masking_utils.py [hunk 4]
```diff
@@ -974,10 +1007,12 @@ def create_causal_mask(
         allow_is_causal_skip = False
         use_vmap = True
 
-    # If we detected packing format
+    # If we detected packing format or blockwise overlay
     if packed_sequence_mask is not None:
         mask_factory_function = and_masks(mask_factory_function, packed_sequence_mask_function(packed_sequence_mask))
         allow_is_causal_skip = False
+    if block_sequence_ids is not None:
+        allow_is_causal_skip = False
```
Comment (Cyrilvallez) [inline src/transformers/masking_utils.py:1020]: See previous message!

File: src/transformers/masking_utils.py [hunk 5]
```diff
@@ -1463,6 +1532,7 @@ def create_masks_for_generate(
         "position_ids": position_ids,
         "or_mask_function": or_mask_function,
         "and_mask_function": and_mask_function,
+        "block_sequence_ids": block_sequence_ids,
```
Comment (Cyrilvallez) [inline src/transformers/masking_utils.py:1529]: Note that `prepare_inputs_for_generation` will never forward this arg, probably it needs to be added!

**Review by Cyrilvallez — COMMENTED** [review]
Nice, I like having this natively!
Just a few things, most particularly I think it's best to pad the mask directly once at the beginning, similar to https://github.com/huggingface/transformers/blob/main/src/transformers/masking_utils.py#L185-L194

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, gemma4, git, paligemma, pi0

**Comment by zucchini-nlp** [comment]
Addressed the comments @Cyrilvallez

File: src/transformers/masking_utils.py [hunk 6]
```diff
@@ -1269,6 +1330,12 @@ def create_bidirectional_sliding_window_mask(
         allow_is_bidirectional_skip = False
         use_vmap = True
 
+    # If we detect a blockwise overlay
+    if block_sequence_ids is not None:
+        block_sequence_ids = maybe_pad_block_sequence_ids(block_sequence_ids, kv_length=kv_length)
+        mask_factory_function = and_masks(mask_factory_function, blockwise_overlay(block_sequence_ids))
+        allow_is_bidirectional_skip = False
```
Comment (vasqu) [inline src/transformers/masking_utils.py]: Why do we need this in the bidirectional masks? Doesn't really make much sense to me tbh 😅

Comment (zucchini-nlp) [reply src/transformers/masking_utils.py]: for consistency in signature only, ig we can not support the arg on purpose? Or consume via kwargs

Comment (vasqu) [reply src/transformers/masking_utils.py]: Imo, we can just let them be consumed by the kwargs :D it just doesn't really make much sense in the context of already bidirectional masks

**Comment by vasqu** [comment]
I think we wait on Cyril for a final pass? I can also take a look if you want but I don't think it's rushing

File: src/transformers/masking_utils.py [hunk 7]
```diff
@@ -194,6 +212,17 @@ def prepare_padding_mask(attention_mask: torch.Tensor | None, kv_length: int, kv
...
 
+def maybe_pad_block_sequence_ids(block_sequence_ids: torch.Tensor, kv_length: int) -> torch.Tensor:
+    """
+    Pads the `block_sequence_ids` in case the total length is less than `kv_length`.
+    Usually that happens with `StaticCache` generation. Pads to the right with `-1`.
+    """
+    if block_sequence_ids.shape[1] < kv_length:
+        to_pad = kv_length - block_sequence_ids.shape[1]
+        block_sequence_ids = F.pad(block_sequence_ids, pad=(0, to_pad), value=-1)
+    return block_sequence_ids
```
Comment (Cyrilvallez) [inline src/transformers/masking_utils.py]: Here it won't pad correctly in the case of sliding window I believe - it should be the exact same as for the padding mask, i.e. the padding length should take the offset into account as well

```python
if (padding_length := kv_length + kv_offset - block_sequence_ids.shape[-1]) > 0:
        block_sequence_ids = F.pad(block_sequence_ids, (0, padding_length), value=-1)
```

**Review by Cyrilvallez — COMMENTED** [review]
I believe we still need to take the offset into account (see comment), but otherwise looks good to me! cc @vasqu here for last check and approval as you checked it as well!

File: src/transformers/masking_utils.py (hunk 7 above)
Comment (zucchini-nlp) [reply src/transformers/masking_utils.py]: ah yeah, we just don't pass it after prefill, so it wasn't complaining. Will update for full correctness

**Comment by zucchini-nlp** [comment]
final pass with slow CI on rebased branch

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, gemma4, git, paligemma, pi0

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, gemma4, git, paligemma, pi0

File: src/transformers/models/gemma3/modular_gemma3.py [hunk 8]
```diff
@@ -703,14 +661,30 @@ def forward(
...
+
+            if token_type_ids is not None:
+                # First find where a new image block starts: 1 if image and previous not image
+                # The images cannot attend to future images, but can attend to all prev images and to itself bidirectionally
+                is_image = (token_type_ids == 1).to(inputs_embeds.device)
+                is_previous_image = nn.functional.pad(is_image, (1, 0), value=0)[:, :-1]
+                new_image_start = is_image & ~is_previous_image
+                group_ids = torch.cumsum(new_image_start.int(), dim=1) - 1
+                group_ids = torch.where(is_image, group_ids, -1)
+                mask_kwargs["block_sequence_ids"] = group_ids
```
Comment (vasqu) [inline src/transformers/models/gemma3/modular_gemma3.py]: Imo, we could still keep these as helper functions. They are identical if I see it correctly (below)

File: src/transformers/models/gemma4/modular_gemma4.py [hunk 9]
```diff
@@ -2110,23 +2067,29 @@ def create_masks_for_generate(
...
-                config, inputs_embeds, attention_mask, past_key_values, position_ids, **kwargs
-            )
+            vision_group_ids = torch.full([*inputs_embeds.size()[:-1]], -1, device=inputs_embeds.device)
+            if mm_token_type_ids is not None:
+                is_vision = (mm_token_type_ids == 1) | (mm_token_type_ids == 2)
+                is_prev_vision = torch.roll(is_vision, shifts=1, dims=-1)
+                is_prev_vision[..., 0] = False
+                new_vision_starts = is_vision & ~is_prev_vision
+                vision_group_ids = torch.cumsum(new_vision_starts.int(), dim=1) - 1
+                vision_group_ids = torch.where(is_vision, vision_group_ids, -1)
```
Comment (vasqu) [inline src/transformers/models/gemma4/modular_gemma4.py]: Same comment re global helper

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 10]
```diff
@@ -353,21 +254,30 @@ def forward(
...
+        if getattr(self.config.text_config, "sliding_window", None) is not None:
+            sliding_mask_kwargs = mask_kwargs.copy()
+            causal_mask = {
+                "full_attention": causal_mask,
+                "sliding_attention": create_sliding_window_causal_mask(**sliding_mask_kwargs),
+            }
```
Comment (vasqu) [inline src/transformers/models/paligemma/modeling_paligemma.py:277]: Maybe you want something like I did in laguna with lazy mask fn creation?

File: src/transformers/models/pi0/configuration_pi0.py [hunk 11]
```diff
@@ -125,8 +125,8 @@ def __post_init__(self, **kwargs):
                 vocab_size=self.vlm_config.text_config.vocab_size,
             )
 
-        # Force bidirectional attention
-        self.dit_config.is_causal = False
+        # Force bidirectional attention for images in Paligemma
+        self.dit_config.is_causal = True
```
Comment (vasqu) [inline src/transformers/models/pi0/configuration_pi0.py:129]: Hmm, so it was wrong before?

File: src/transformers/models/pi0/modular_pi0.py [hunk 12]
```diff
@@ -607,13 +612,16 @@ def sample_actions(
             )
 
         # 2. Run VLM once and obtain prefix cache. Must infer positions here!
+        position_ids = None
         if attention_mask is not None:
             position_ids = attention_mask.cumsum(-1) - 1
         inputs_embeds = self.model.embed_prefix(input_ids, pixel_values, pixel_attention_mask)
+        token_type_ids = torch.zeros_like(inputs_embeds)[:, :, 0]
```
Comment (vasqu) [inline src/transformers/models/pi0/modular_pi0.py:620]: Let's add a comment for clarification

**Review by vasqu — APPROVED** [review]
Not much to complain tbh, let's check for the last details and run-slow to sanity check 🫡

File: src/transformers/models/pi0/configuration_pi0.py (hunk 11 above)
Comment (zucchini-nlp) [reply src/transformers/models/pi0/configuration_pi0.py:129]: nah, I was using a `create_biditrectional_mask` API before which I thought fits more to get a blockwise causal mask 🥲 

It has to be causal if we want to use the `block_sequence_ids`, and the ckpt on the hub isn't converted yet anyway. No prob with BC

File: src/transformers/models/gemma3/modular_gemma3.py (hunk 8 above)
Comment (zucchini-nlp) [reply src/transformers/models/gemma3/modular_gemma3.py]: Not always but for the two gemma models, it is same. Can split out to fn indeed

update: oh actually not same, each model uses its own way to map type-ids to `block-sequences`. Gemma3 has no videos, PG splits with prefix/suffix and PI0 is a completely different model since it is a blockwise DiT

I'll split within each gemma file, so we don't repeat twice

**Comment by zucchini-nlp** [comment]
Thanks I will fix and try to merge on Monday, wanted to concentrate on Kimi today

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 10 above)
Comment (zucchini-nlp) [reply src/transformers/models/paligemma/modeling_paligemma.py:277]: looked at it just now. Not sure it fits well, because PG uses a completely different config for two ckpt, so we don't always have attr `layer_types` and same about `config.sliding_window`

File: src/transformers/models/pi0/modular_pi0.py (hunk 12 above)
Comment (zucchini-nlp) [reply src/transformers/models/pi0/modular_pi0.py:620]: oke, but it is same thing as in `forward`. I just forgot to add it when sampling, and luckily it was falling back to correct token types

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, gemma4, git, paligemma, pi0

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, gemma4, git, paligemma, pi0

**Comment by zucchini-nlp** [comment]
run-slow: pi0

**Comment by zucchini-nlp** [comment]
Failing test is an OOM and already fails on `main`. Merging
